pds_doi_core/util/doi_validator.py

```diff
@@ -0,0 +1,323 @@
+            col_index_to_lid     = columns.index('lid')
+            col_index_to_vid     = columns.index('vid')
+            col_index_to_status  = columns.index('status')
+            logger.debug(f"action: {action} rows[last_row_index][col_index_to_doi] {rows[last_row_index][col_index_to_doi]}")
+            lidvid = rows[last_row_index][col_index_to_lid] + '::' + rows[last_row_index][col_index_to_vid]
+            if rows[last_row_index][col_index_to_status].lower() not in self.m_workflow_order:
+                logger.error(f"Unexpected status of doi {rows[last_row_index][col_index_to_status].lower()} from database.  Valid values are {self.m_workflow_order.keys()}")
+                raise InvalidDOIException(f"Unexpected status of doi {rows[last_row_index][col_index_to_status].lower()} from database.  Valid values are {self.m_workflow_order.keys()}")
+
+            # Translate the 'status' field into a ranking in the workflow, e.g. 'Draft' to 2
+            rank_current_doi_status = self.m_workflow_order[doi.status.lower()]
+
+            # Translate the 'status' field of same lidvid in database into a ranking in the workflow, e.g. 'Pending' to 3
+            rank_database_doi_status = self.m_workflow_order[rows[last_row_index][col_index_to_status].lower()]
+
+            logger.debug(f"rank_database_doi_status,rank_current_doi_status {rank_database_doi_status,rank_current_doi_status}")
+
+            # A status tuple of ('Pending',3) is higher than ('Draft',2) will cause an error unless --force flag is used.
+            if rank_database_doi_status > rank_current_doi_status:
+                logger.debug(f"rows[last_row_index][col_index_to_status].lower(),doi.status.lower() {rows[last_row_index][col_index_to_status].lower(),doi.status.lower()}")
+                logger.debug(f"CHECK_FIELD_WORKFLOW_GOOD_FALSE: lidvid: {lidvid} status: {doi.status.lower()}")
+                logger.error(f"There is a DOI: {rows[last_row_index][col_index_to_doi]} record which status: '{rows[last_row_index][col_index_to_status].lower()}'.  Are you sure you want to restart the workflow from step: {rank_current_doi_status} for the lidvid: {doi.related_identifier}?")
+                # If the flag --force is set do not throw an exception, otherwise throw an exception to leave the code.
+
+                if force_flag:
+                    pass
+                else:
+                    raise UnexpectedDOIActionException(f"There is a DOI: {rows[last_row_index][col_index_to_doi]} record which status: '{rows[last_row_index][col_index_to_status].lower()}'.  Are you sure you want to restart the workflow from step: {rank_current_doi_status} for the lidvid: {doi.related_identifier}?")
+            else:
+                logger.debug(f"CHECK_FIELD_WORKFLOW_GOOD_TRUE: lidvid: {lidvid} status: {doi.status.lower()} status (db): {rows[last_row_index][col_index_to_status].lower()}") 
+
+        return 1
+
+    def validate(self, doi: Doi, action, force_flag=False):
```

Comment:
This method looks good, easy to read.
To make it even more concise, I think you can move the detail comments on each rule to their respective method. The method name are self explanatory at this level.

Reply:
Can you also check if we really need the action argument ?

Reply:
I am pretty sure the action is needed, especially for these two function:

        self._check_field_title_duplicate(doi,action)
        self._check_field_lidvid_update(doi,action)

The ones that don't really need them have been removed.

when the _check_field_lidvid_update() function is called.